User improvements (#612)
foundriesio/docs: source/user-guide/troubleshooting/troubleshooting.rst

  fioctl targets prune --by-tag <tag> --keep-last <number> --dryrun

Device Registration Common Errors
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Even if the device has a proper internet connection, users can still run into errors during device registration.
The ``lmp-device-register`` usually provides some diagnostics in the error message.
However in certain cases it does not expose sensitive information to avoid possible attack vectors.

> Common errors encoutered during device registration are:

Reviewer: encountered

  fioctl targets prune --by-tag <tag> --keep-last <number> --dryrun

Device Registration Common Errors
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Even if the device has a proper internet connection, users can still run into errors during device registration. ``lmp-device-register`` usually provides some diagnostics in the error message, however in certain cases it does not expose enough information to avoid possible attack vectors.

Reviewer 2: ```suggestion
Even if the device has a proper internet connection, you may still run into errors during device registration. ``lmp-device-register`` usually provides some diagnostics in the error message, however in certain cases it does not expose enough information to avoid possible attack vectors.
```

Common errors encoutered during device registration are:

* **Error 401**

.. code-block::

   Unable to create device: HTTP_401

This indicates a problem with the token.

**Solution:** Verify there is a valid non-expired token in https://app.foundries.io/settings/tokens/.

* **Error 403**

.. code-block::

   Unable to create device: HTTP_403

This indicates no permission to create a device in the Factory.

**Solution:** Verify the user token has ``device:create`` scope in https://app.foundries.io/settings/tokens/. If the Factory has :ref:`ref-team-based-access` set, check if the user is part of a team which has ``device:create`` permissions.

> * **Wrong Factory value**

Reviewer 3: What is the status code in this case? I guess 403, right?

I think for these 3 sections we need to choose what we show: deliberate error category name, API error, or `lmp-device-register` message.

From the user perspective, a message would be the best, but it might look weird as a section name.
So, maybe, we show a deliberate error category name (like you did here) for all sections, and then the exact error message goes below in a `code-block`, just like you did.

The idea here is to have a uniform look-and-feel, which is easy to extend.

Author: For Wrong Factory value this is what I see:
`Token is not specified, getting an oauth token from Foundries' auth endpoint...

----------------------------------------------------------------------------
Visit the link below in your browser to authorize this new device. This link
will expire in 15 minutes.
  Device Name: 505ae417-4a7c-4b90-a9db-0ac279e4d0fb
  User code: 32e4-3858
  Browser URL: https://app.foundries.io/activate/

Error authorizing device: 'scope' parameter is not valid: wrong Factory value, or you are not a member of the Factory`

While 403:
`Token is not specified, getting an oauth token from Foundries' auth endpoint...

----------------------------------------------------------------------------
Visit the link below in your browser to authorize this new device. This link
will expire in 15 minutes.
  Device Name: e312da04-6ac5-4ed5-a4ee-d375bbe42b66
  User code: cca3-577a
  Browser URL: https://app.foundries.io/activate/

Registering device, test-homehub-809, to factory home-hub.
Device UUID: e312da04-6ac5-4ed5-a4ee-d375bbe42b66
Unable to create device: HTTP_403
message: A factory admin must add you to a team with one of these scopes: home-hub:devices:create`

And 401:
`Device UUID: 0ddc1789-046f-475a-8a81-a9817ebce4cd
Unable to create device: HTTP_401
Polis Error: {"error":"not_found","error_description":"Cannot find a user with the provided token","status":404}`

Author: They are not very consistent but I can put the whole message error in the page for clarity.

Author: I simplified by only showing the message error.

.. code-block::

   Error authorizing device: 'scope' parameter is not valid: wrong Factory value

> This usually means the device is running an image which was built locally and not on FoundriesFactory CI.

Reviewer 3: Is this the only use case?

I guess when we build in CI, the factory name is obtained from `/etc/os-release`, right?

I know at least 4 use cases when this is not true:
- A local build without specifying the factory name as a variable.
- An LmP factory.
- A user has given a token for the wrong factory as an argument.
- A user has given a wrong factory name as an argument.

The third use case above is important, as we recommend to not use auto-register for production devices.
Hence, a user needs to either provide a token or use Oauth2 flow.

@ldts @mike-sul do you know other use cases?

Author: The most frequent (if not only) case where we get this error complaint from customers is when they try to register from local builds, so maybe we don't need to go further here.

**Solution:** Flash an image built from CI.

> 

Author: @vkhoroz Could you please review this section?

  fioctl targets prune --by-tag <tag> --keep-last <number> --dryrun

Device Registration Common Errors
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Even if the device has a proper internet connection, users can still run into errors during device registration.
The ``lmp-device-register`` usually provides some diagnostics in the error message.

> However in certain cases it does not expose sensitive information to avoid possible attack vectors.

Reviewer 4: I don't quite understand line 262. Why is it negative that the log error does not have sensitive information to avoid possible attacks? Wouldn't it be a good and desirable thing?

Author: It's not a bad thing, it's indeed a desirable thing, this is the reason we don't show it in the logs.

Reviewer 4: In this case, I would not use "however". I would use something as "and", or "so". Perhaps:

The ``lmp-device-register`` usually provides some diagnostics in the error message without exposing sensitive information to avoid possible attack vectors, in some cases.

Does it mean in some cases it does expose sensitive information?

Reviewer 3: For the 401 error we usually show very opaque (limited) error message, so that no sensitive information can be leaked to malicious actors..
That is desired, but some users complain that they don't get enough diagnostics from it.

I believe this is the primary reason why @vanmaegima created this troubleshooting page.

I personally like the direction @angolini is heading towards.
I think we need to elaborate that way somehow.

Maybe, we state it explicitly that we always try to show proper diagnostics to the user.
For example, if error happens during the authorization phase (403 error) - there is a plenty of information in the error message, so that the user can easily deduct what permissions they are missing to perform an action.

But the authentication use case is very special. Any hint we pass with the 401 error could help the potential attacker to perform forgery of sensitive information, including personal user information, or clues how to overcome the passwall. So, maybe we state it explicitly that in this case an opaque useless message is for a reason.
Then this troubleshooting page might help legitimate users to see what would be their steps to find a way in.

IMHO I wrote too much detail above; so it is definitely not concise enough for the user manual.
But I'm more than sure that you guys will be able to conduct proper messaging based on my braindump.

Common errors encoutered during device registration are:

* **Error 401**

.. code-block::

   Unable to create device: HTTP_401

This indicates a problem with the token.

**Solution:** Verify there is a valid non-expired token in https://app.foundries.io/settings/tokens/.

* **Error 403**

.. code-block::

   Unable to create device: HTTP_403

This indicates no permission to create a device in the Factory.

> **Solution:** Verify the user token has ``device:create`` scope in https://app.foundries.io/settings/tokens/. If the Factory has :ref:`ref-team-based-access` set, check if the user is part of a team which has ``device:create`` permissions.

Reviewer 4: I prefer to use semantic break lines.

4. Lastly, perform the registration again.

.. _ref-ts-fiovb-container:

Read Secure Variables from Containers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> After a board is fused and closed, the secure storage (RPMB) becomes available and handles the necessary variables to perform the OTA logic. Secure storage also can be leveraged to store custom device information, like MAC addresses, serial numbers and other relevant values.

Reviewer 2: ```suggestion
After a board is fused and closed, the secure storage (RPMB) becomes available and handles the necessary variables to perform the OTA logic. Secure storage also can be leveraged to store custom device information, like MAC addresses, serial numbers, and other relevant values.
```

> It is common that users would like to retrieve these values from the application. For that purpose, please refer to the `fiovb-container <https://github.com/foundriesio/containers/tree/master/fiovb-container>`_ example, which brings a simple application to run ``fiovb_printenv`` from inside a container.

Reviewer 2: ```suggestion
You may wish to retrieve these values from the application. Please refer to the `fiovb-container <https://github.com/foundriesio/containers/tree/master/fiovb-container>`_ example, which brings a simple application to run ``fiovb_printenv`` from inside a container.
```

4. Lastly, perform the registration again.

.. _ref-ts-fiovb-container:

Read Secure Variables from Containers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> After a board is fused and closed, the secure storage (RPMB) becomes available and handles the necessary variables to perform the OTA logic. Secure storage also can be leveraged to store custom device information, like MAC addresses, serial numbers, and other relevant values.

Reviewer 4: I prefer to use semantic break lines.
